File: src/main/java/hiiragi283/material/util/CraftingBuilder.java
package hiiragi283.material.util;

import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.Ingredient;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.common.registry.GameRegistry;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.ArrayList;
import java.util.List;

@ParametersAreNonnullByDefault
public class CraftingBuilder {

    private final ResourceLocation location;
    private final ItemStack output;

    public CraftingBuilder(ItemStack output) {
        this(HiiragiUtil.toLocation(output, '_'), output);
    }

    public CraftingBuilder(String location, ItemStack output) {
        this(new ResourceLocation(location), output);
    }

    public CraftingBuilder(ResourceLocation location, ItemStack output) {
        this.location = location;
        this.output = output;
    }


    //    Shaped    //

    private final List<Object> PARAM_SHAPED = new ArrayList<>();

    public CraftingBuilder setPattern(String... patterns) {
        PARAM_SHAPED.add(patterns[0]);
        PARAM_SHAPED.add(patterns[1]);
        PARAM_SHAPED.add(patterns[2]);
        return this;
    }

    public CraftingBuilder setIngredient(Character mark, ItemStack input) {
        PARAM_SHAPED.add(mark);
        PARAM_SHAPED.add(input);
        return this;
    }

    public CraftingBuilder setIngredient(Character mark, String input) {
        PARAM_SHAPED.add(mark);
        PARAM_SHAPED.add(input);
        return this;
    }

    //    Shapeless    //

    private final List<Ingredient> PARAM_SHAPELESS = new ArrayList<>();

    public CraftingBuilder addIngredient(Ingredient ingredient) {
        PARAM_SHAPELESS.add(ingredient);
        return this;
    }

    //    Build    //

    public void build() {
        if (!PARAM_SHAPED.isEmpty()) {
            GameRegistry.addShapedRecipe(location, null, output, PARAM_SHAPED.toArray());
        } else if (!PARAM_SHAPELESS.isEmpty()) {
            GameRegistry.addShapelessRecipe(location, null, output, PARAM_SHAPELESS.toArray(new Ingredient[0]));
        } else {
            throw new IllegalStateException("Either shaped parameters or shapeless parameters should be empty!");
        }
    }

}

File: src/main/kotlin/hiiragi283/material/util/CraftingBuilder.kt
package hiiragi283.material.util

import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.Ingredient
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.crafting.CraftingHelper
import net.minecraftforge.fml.common.registry.GameRegistry

class CraftingBuilder(private val location: ResourceLocation, private val output: ItemStack) {

    constructor(location: String, output: ItemStack) : this(ResourceLocation(location), output)

    constructor(output: ItemStack, alt: String = "") : this(output.toLocation("_").append(alt), output)

    fun build() {
        when {
            params.isNotEmpty() -> GameRegistry.addShapedRecipe(location, null, output, *params.toTypedArray())
            ingredients.isNotEmpty() -> GameRegistry.addShapelessRecipe(location, null, output, *ingredients.toTypedArray())
            else -> throw IllegalStateException("Either shaped parameters or shapeless parameters should be empty!")
        }
    }

    //    Shaped    //

    private val params: MutableList<Any> = mutableListOf()

    fun setPattern(vararg pattern: String) = also { params.addAll(pattern) }

    fun setIngredient(mark: Char, input: ItemStack) = also {
        params.add(mark)
        params.add(input)
    }

    fun setIngredient(mark: Char, input: Block, isWild: Boolean = false) =
        setIngredient(mark, ItemStack(input, 1, if (isWild) Short.MAX_VALUE.toInt() else 0))

    fun setIngredient(mark: Char, input: Item, isWild: Boolean = false) =
        setIngredient(mark, ItemStack(input, 1, if (isWild) Short.MAX_VALUE.toInt() else 0))

    fun setIngredient(mark: Char, oredict: String) = also {
        params.add(mark)
        params.add(oredict)
    }

    //    Shapeless    //

    private val ingredients: MutableList<Ingredient> = mutableListOf()

    fun addIngredient(obj: Any) = also {
        ingredients.add(CraftingHelper.getIngredient(obj))
    }

}